### PR TITLE
Add check for cluster_scoped_gc

### PR DESCRIPTION
**Pull request by daxmc99**
Add check for cluster scoped finalizers

Problem: Cluster scoped finalizers were preventing deletion of
    resources. The previous solution of manually writing cleanup code
    within cluster_scoped_gc.go was difficult to maintain.

Solution: All generated handlers have an init function that adds to a
    map within the norman framework. During cluster deletion this map is
    iterated over and finalizers are removed as needed.

Addresses #19550

_Dependency on_:

1. https://github.com/rancher/norman/pull/277
2. https://github.com/rancher/types/pull/798

### CODE REVIEW

**Comment by daxmc99** [comment]
- [ ] Ensure that nodepool & node controller cleans up after themselves without any functionality within cluster_scoped_gc for clusters with >1 nodes.

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 1]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
...
+		for _, obj := range objList.Items {
+			_, err = cleanFinalizersGeneric(cluster.Name, &obj, c.mgmt.DynamicClient.Resource(value).Namespace(obj.GetNamespace()))
+			if err != nil {
+				return cluster, err
+			}
+		}
+	}
+	resource.CleanerFuncMap.RUnlock()
+
 	rts, err := c.rtLister.List("", labels.Everything())
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: remove legacy cleaner uppers

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 2]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
 
 	if modified {
 		md.SetFinalizers(finalizers)
-		_, err := client.Update(md.GetName(), object)
+		//_, err := client.Update(md.GetName(), object)
+		// we don't remove all the nodes in generic... does that matter? Seems like cleanup might occur somewhere else,
+		// didn't see any stuck finalizers
+		logrus.Warnln("CLEAN GENERIC MISSED THIS OBJECT ", object)
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: remove this stuff

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 3]
```diff
@@ -2,6 +2,13 @@ package clustergc
 
 import (
 	"context"
+	"fmt"
+	"github.com/rancher/norman/resource"
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: separate out standard libs from vendored stuff

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 4]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
...
+		//_, err := client.Update(md.GetName(), object)
+		// we don't remove all the nodes in generic... does that matter? Seems like cleanup might occur somewhere else,
+		// didn't see any stuck finalizers
+		logrus.Warnln("CLEAN GENERIC MISSED THIS OBJECT ", object)
 		return err
 	}
 	return nil
 }
 
+func cleanFinalizersGeneric(clusterName string, object *unstructured.Unstructured, dynamicClient dynamic.ResourceInterface) (*unstructured.Unstructured, error) {
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: as part of cleanup, get rid of cleanFinalizers and rename this to that

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 5]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
...
+		if strings.HasPrefix(f, lifecycle.ScopedFinalizerKey) && strings.HasSuffix(f, "_"+clusterName) {
+			finalizers = append(finalizers[:i], finalizers[i+1:]...)
+			modified = true
+		}
+	}
+
+	if modified {
+		md.SetFinalizers(finalizers)
+		obj, ok := dynamicClient.Update(object, metav1.UpdateOptions{})
+		fmt.Printf("Removed finalizer from gvr %s \n", object.GetName())
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: remove

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 6]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
...
+		f := finalizers[i]
+		if strings.HasPrefix(f, lifecycle.ScopedFinalizerKey) && strings.HasSuffix(f, "_"+clusterName) {
+			finalizers = append(finalizers[:i], finalizers[i+1:]...)
+			modified = true
+		}
+	}
+
+	if modified {
+		md.SetFinalizers(finalizers)
+		obj, ok := dynamicClient.Update(object, metav1.UpdateOptions{})
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: dont use ok. use e or err2

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go [hunk 7]
```diff
@@ -85,13 +92,63 @@ func cleanFinalizers(clusterName string, object runtime.Object, client *objectcl
...
+		fmt.Printf("Removed finalizer from gvr %s \n", object.GetName())
+		return obj, ok
+	}
+	return object, nil
+}
+
 func (c *gcLifecycle) Remove(cluster *v3.Cluster) (runtime.Object, error) {
+
+	resource.CleanerFuncMap.RLock()
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc.go]: once refactored, wont need this, bc we'll be returning a list

File: pkg/controllers/management/clustergc/cluster_scoped_gc_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"github.com/rancher/norman/lifecycle"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"testing"
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc_test.go]: separate imports

File: pkg/controllers/management/clustergc/cluster_scoped_gc_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+			"DontRemoveUnrelated",
+			"a",
+			finalizerFactory("testObj",
+
+				lifecycle.ScopedFinalizerKey+"App"+"_"+"b",
+				lifecycle.ScopedFinalizerKey+"App"+"_"+"a",
+			),
+			[]string{lifecycle.ScopedFinalizerKey + "App" + "_" + "b"},
+		},
+		{
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc_test.go]: add one more test case for non-scoped finalizers being left untouched

File: pkg/controllers/management/clustergc/cluster_scoped_gc_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+		return err
+	}
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	u.Object["metadata"] = metadata
+	return nil
+}
+
+type mockInterface struct{}
```
Comment (cjellick) [inline pkg/controllers/management/clustergc/cluster_scoped_gc_test.go]: rename mockDynamicResourceInterface

File: pkg/controllers/management/clustergc/cluster_scoped_gc.go (hunk 5 above)
Comment (daxmc99) [reply pkg/controllers/management/clustergc/cluster_scoped_gc.go]: Changed to a debugf

**Comment by cjellick** [comment]
LGTM

**Pull request merged**